## Импортируем необходимые библиотеки

In [1]:
import requests
from bs4 import BeautifulSoup
import lxml
import json
import pandas as pd

Еще в прошлом году работало прекрасно API по NHL - statsapi.web.nhl.com  

Поймем какую информацию можно получить о хоккейных матчах

Здесь и ниже 8471214 - id Александра Овечкина

In [3]:
url = 'https://statsapi.web.nhl.com/api/v1/people/8471214/stats/?stats=statsSingleSeason&season=20212022'
response = requests.get(url=url)
response.encoding = 'utf-8'
soup = BeautifulSoup(response.text, 'lxml')
soup

<html><body><p>{
  "copyright" : "NHL and the NHL Shield are registered trademarks of the National Hockey League. NHL and NHL team marks are the property of the NHL and its teams. © NHL 2023. All Rights Reserved.",
  "stats" : [ {
    "type" : {
      "displayName" : "statsSingleSeason",
      "gameType" : {
        "id" : "R",
        "description" : "Regular season",
        "postseason" : false
      }
    },
    "splits" : [ {
      "season" : "20212022",
      "stat" : {
        "timeOnIce" : "1584:02",
        "assists" : 40,
        "goals" : 50,
        "pim" : 18,
        "shots" : 334,
        "games" : 77,
        "hits" : 155,
        "powerPlayGoals" : 16,
        "powerPlayPoints" : 29,
        "powerPlayTimeOnIce" : "366:21",
        "evenTimeOnIce" : "1216:50",
        "penaltyMinutes" : "18",
        "faceOffPct" : 30.76,
        "shotPct" : 15.0,
        "gameWinningGoals" : 5,
        "overTimeGoals" : 0,
        "shortHandedGoals" : 1,
        "shortHandedPoints" : 

In [4]:
url = 'https://statsapi.web.nhl.com/api/v1/people/8471214/stats/?stats=statsSingleSeason&season=20212022'
response = requests.get(url=url).json()
response

{'copyright': 'NHL and the NHL Shield are registered trademarks of the National Hockey League. NHL and NHL team marks are the property of the NHL and its teams. © NHL 2023. All Rights Reserved.',
 'stats': [{'type': {'displayName': 'statsSingleSeason',
    'gameType': {'id': 'R',
     'description': 'Regular season',
     'postseason': False}},
   'splits': [{'season': '20212022',
     'stat': {'timeOnIce': '1584:02',
      'assists': 40,
      'goals': 50,
      'pim': 18,
      'shots': 334,
      'games': 77,
      'hits': 155,
      'powerPlayGoals': 16,
      'powerPlayPoints': 29,
      'powerPlayTimeOnIce': '366:21',
      'evenTimeOnIce': '1216:50',
      'penaltyMinutes': '18',
      'faceOffPct': 30.76,
      'shotPct': 15.0,
      'gameWinningGoals': 5,
      'overTimeGoals': 0,
      'shortHandedGoals': 1,
      'shortHandedPoints': 1,
      'shortHandedTimeOnIce': '00:51',
      'blocked': 25,
      'plusMinus': 8,
      'points': 90,
      'shifts': 1680,
      'timeOnIce

Видим что по результату GET-запроса получаем json файл со статистикой игрока за сезон

In [5]:
response.keys()

dict_keys(['copyright', 'stats'])

In [6]:
response['stats'][0]

{'type': {'displayName': 'statsSingleSeason',
  'gameType': {'id': 'R',
   'description': 'Regular season',
   'postseason': False}},
 'splits': [{'season': '20212022',
   'stat': {'timeOnIce': '1584:02',
    'assists': 40,
    'goals': 50,
    'pim': 18,
    'shots': 334,
    'games': 77,
    'hits': 155,
    'powerPlayGoals': 16,
    'powerPlayPoints': 29,
    'powerPlayTimeOnIce': '366:21',
    'evenTimeOnIce': '1216:50',
    'penaltyMinutes': '18',
    'faceOffPct': 30.76,
    'shotPct': 15.0,
    'gameWinningGoals': 5,
    'overTimeGoals': 0,
    'shortHandedGoals': 1,
    'shortHandedPoints': 1,
    'shortHandedTimeOnIce': '00:51',
    'blocked': 25,
    'plusMinus': 8,
    'points': 90,
    'shifts': 1680,
    'timeOnIcePerGame': '20:34',
    'evenTimeOnIcePerGame': '15:48',
    'shortHandedTimeOnIcePerGame': '00:00',
    'powerPlayTimeOnIcePerGame': '04:45'}}]}

In [7]:
import dpath
dpath.get(response, '**/goals')

50

## Получим более детальную статистику по каждой игре в сезоне 16-17

In [8]:
url = 'https://statsapi.web.nhl.com/api/v1/people/8471214/stats/?stats=gameLog&season=20162017'
stats16_17 = requests.get(url=url).json()
stats16_17.keys()

dict_keys(['copyright', 'stats'])

In [9]:
games16_17 = stats16_17['stats'][0]['splits']

In [10]:
games16_17[0]

{'season': '20162017',
 'stat': {'timeOnIce': '17:56',
  'assists': 0,
  'goals': 0,
  'pim': 0,
  'shots': 5,
  'games': 1,
  'hits': 0,
  'powerPlayGoals': 0,
  'powerPlayPoints': 0,
  'powerPlayTimeOnIce': '02:44',
  'evenTimeOnIce': '15:12',
  'penaltyMinutes': '0',
  'shotPct': 0.0,
  'gameWinningGoals': 0,
  'overTimeGoals': 0,
  'shortHandedGoals': 0,
  'shortHandedPoints': 0,
  'shortHandedTimeOnIce': '00:00',
  'blocked': 0,
  'plusMinus': -1,
  'points': 0,
  'shifts': 20},
 'team': {'id': 15, 'name': 'Washington Capitals', 'link': '/api/v1/teams/15'},
 'opponent': {'id': 13,
  'name': 'Florida Panthers',
  'link': '/api/v1/teams/13'},
 'date': '2017-04-09',
 'isHome': True,
 'isWin': False,
 'isOT': False,
 'game': {'gamePk': 2016021228,
  'link': '/api/v1/game/2016021228/feed/live',
  'content': {'link': '/api/v1/game/2016021228/content'}}}

Получили информацию по первой игре сезона, посмотрим какие параметры она содержит

In [11]:
games16_17[0].keys()

dict_keys(['season', 'stat', 'team', 'opponent', 'date', 'isHome', 'isWin', 'isOT', 'game'])

In [12]:
 games16_17[0]['isHome']

True

In [13]:
game_id = games16_17[0]['game']['gamePk']
game_id

2016021228

In [14]:
game_date = games16_17[0]['date']
game_date

'2017-04-09'

In [17]:
games16_17[0]['opponent']['name']

'Florida Panthers'

In [18]:
game_1 = pd.DataFrame.from_dict(games16_17[0]).T

In [19]:
ff = games16_17[0]['stat']
ff = pd.DataFrame.from_dict(ff, orient='index')
ff

0
timeOnIce             17:56
assists                   0
goals                     0
pim                       0
shots                     5
games                     1
hits                      0
powerPlayGoals            0
powerPlayPoints           0
powerPlayTimeOnIce    02:44
evenTimeOnIce         15:12
penaltyMinutes            0
shotPct                 0.0
gameWinningGoals          0
overTimeGoals             0
shortHandedGoals          0
shortHandedPoints         0
shortHandedTimeOnIce  00:00
blocked                   0
plusMinus                -1
points                    0
shifts                   20

In [20]:
new_index = ['Florida Panthers']

ff['Florida Panthers'] = ff[0]

In [21]:
ff = ff.T

In [22]:
ff['date'] = game_date
ff['opponent'] = 'Florida Panthers'
ff

timeOnIce assists goals pim shots games hits powerPlayGoals  \
0                    17:56       0     0   0     5     1    0              0   
Florida Panthers     17:56       0     0   0     5     1    0              0   

                 powerPlayPoints powerPlayTimeOnIce  ... overTimeGoals  \
0                              0              02:44  ...             0   
Florida Panthers               0              02:44  ...             0   

                 shortHandedGoals shortHandedPoints shortHandedTimeOnIce  \
0                               0                 0                00:00   
Florida Panthers                0                 0                00:00   

                 blocked plusMinus points shifts        date          opponent  
0                      0        -1      0     20  2017-04-09  Florida Panthers  
Florida Panthers       0        -1      0     20  2017-04-09  Florida Panthers  

[2 rows x 24 columns]

In [23]:
games16_17[0].keys()

dict_keys(['season', 'stat', 'team', 'opponent', 'date', 'isHome', 'isWin', 'isOT', 'game'])

In [24]:
games16_17[0]['season']

'20162017'

In [25]:
order = ['date', 'opponent', 'goals', 'assists', 'points', 'plusMinus', 'powerPlayGoals', 'powerPlayPoints','powerPlayTimeOnIce','shortHandedGoals', 'gameWinningGoals',
       'overTimeGoals', 'shots', 'penaltyMinutes', 'shifts', 'timeOnIce', 'isHome', 'isWin', 'isOT', 'season']

In [26]:
def get_game_info(game):
    df = game['stat']
    df = pd.DataFrame.from_dict(df, orient='index')
    game_id = game['game']['gamePk']
    df = df.T
    df.index = [game_id]
    df['date'] = game['date']
    df['opponent'] = game['opponent']['name']
    df['isHome'] = game['isHome']
    df['isWin'] = game['isWin']
    df['isOT'] = game['isOT']
    df['season'] = game['season']
    df['date']=pd.to_datetime(df['date'], format='%Y-%m-%d')
    df['timeOnIce']=pd.to_datetime(df['timeOnIce'], format='%M:%S')
    df['powerPlayTimeOnIce']=pd.to_datetime(df['powerPlayTimeOnIce'], format='%M:%S')
    df = df[order]
    return df

In [27]:
get_game_info(games16_17[0])

date          opponent goals assists points plusMinus  \
2016021228 2017-04-09  Florida Panthers     0       0      0        -1   

           powerPlayGoals powerPlayPoints powerPlayTimeOnIce shortHandedGoals  \
2016021228              0               0              02:44                0   

           gameWinningGoals overTimeGoals shots penaltyMinutes shifts  \
2016021228                0             0     5              0     20   

           timeOnIce  isHome  isWin   isOT    season  
2016021228     17:56    True  False  False  20162017

Немного поигрались с формой данных, создав функцию получения информации об одной игре  
Теперь нужнно получить все игры всех сезонов

# Пытаемся собрать все игры, все сезоны (на момент создания файла)

In [29]:
start_hnl = '2005'
base_URL = 'https://statsapi.web.nhl.com/api/v1/people/8471214/stats/?stats=gameLog&season='
seasons = []
while start_hnl!='2023':
    season = start_hnl + str(int(start_hnl)+1)
    seasons.append(season)
    start_hnl = str(int(start_hnl)+1)
print(seasons)    

['20052006', '20062007', '20072008', '20082009', '20092010', '20102011', '20112012', '20122013', '20132014', '20142015', '20152016', '20162017', '20172018', '20182019', '20192020', '20202021', '20212022', '20222023']


Формируем список сезонов, чтобы автоматически подставить в ссылку и затем в функцию

In [31]:
total_df = pd.DataFrame()
for i in range(len(seasons)):
    url = 'https://statsapi.web.nhl.com/api/v1/people/8471214/stats/?stats=gameLog&season='+seasons[i]
    stats = requests.get(url=url).json()
    games = stats['stats'][0]['splits']
    for i in range(len(games)):
        df = get_game_info(games[i])
        total_df = pd.concat([total_df,df], axis=0)
total_df

date             opponent goals assists points plusMinus  \
2005021228 2006-04-18  Tampa Bay Lightning     0       1      1         0   
2005021215 2006-04-17    Atlanta Thrashers     1       1      2        -1   
2005021207 2006-04-15     Florida Panthers     1       0      1         1   
2005021185 2006-04-13    Atlanta Thrashers     1       1      2         0   
2005021165 2006-04-10        Boston Bruins     1       0      1         1   
...               ...                  ...   ...     ...    ...       ...   
2022020065 2022-10-20      Ottawa Senators     0       0      0        -2   
2022020043 2022-10-17    Vancouver Canucks     2       2      4         1   
2022020029 2022-10-15   Montréal Canadiens     0       1      1         0   
2022020015 2022-10-13  Toronto Maple Leafs     0       0      0        -1   
2022020005 2022-10-12        Boston Bruins     0       0      0        -2   

           powerPlayGoals powerPlayPoints powerPlayTimeOnIce shortHandedGoals  \
2005021228              0               1              04:15                0   
2005021215              0               1              07:40                0   
2005021207              0               0              05:57                0   
2005021185              0               1              08:42                0   
2005021165              0               0              06:14                0   
...                   ...             ...                ...              ...   
2022020065              0               0              04:51                0   
2022020043              1               1              02:30                0   
2022020029              0               1              03:07                0   
2022020015              0               0              08:30                0   
2022020005              0               0              06:27                0   

           gameWinningGoals overTimeGoals shots penaltyMinutes shifts  \
2005021228                0             0     4              0     18   
2005021215                0             0     6              0     24   
2005021207                0             0     5              0     25   
2005021185                0             0     3              0     23   
2005021165                1             1     4              0     28   
...                     ...           ...   ...            ...    ...   
2022020065                0             0     2              2     21   
2022020043                0             0     4              0     25   
2022020029                0             0     6              0     20   
2022020015                0             0     4              0     19   
2022020005                0             0     3              0     21   

           timeOnIce  isHome  isWin   isOT    season  
2005021228     17:31   False   True  False  20052006  
2005021215     20:09    True   True  False  20052006  
2005021207     25:50   False   True  False  20052006  
2005021185     23:42   False  False  False  20052006  
2005021165     23:20   False   True  False  20052006  
...              ...     ...    ...    ...       ...  
2022020065     19:59   False  False  False  20222023  
2022020043     18:02    True   True  False  20222023  
2022020029     17:27    True   True  False  20222023  
2022020015     22:18   False  False  False  20222023  
2022020005     22:08    True  False  False  20222023  

[1316 rows x 20 columns]

In [ ]:
total_df.info()

In [32]:
total_df = total_df.sort_values(by=['date'])

In [33]:
total_df[total_df['season']=='20162017']

date               opponent goals assists points plusMinus  \
2016020007 2016-10-13    Pittsburgh Penguins     0       0      0        -1   
2016020022 2016-10-15     New York Islanders     0       0      0         0   
2016020039 2016-10-18     Colorado Avalanche     1       1      2         1   
2016020056 2016-10-20       Florida Panthers     1       0      1         2   
2016020071 2016-10-22       New York Rangers     1       0      1         0   
...               ...                    ...   ...     ...    ...       ...   
2016021169 2017-04-02  Columbus Blue Jackets     0       0      0         0   
2016021182 2017-04-04    Toronto Maple Leafs     0       1      1         0   
2016021193 2017-04-05       New York Rangers     0       1      1         0   
2016021209 2017-04-08          Boston Bruins     0       1      1         1   
2016021228 2017-04-09       Florida Panthers     0       0      0        -1   

           powerPlayGoals powerPlayPoints powerPlayTimeOnIce shortHandedGoals  \
2016020007              0               0              04:31                0   
2016020022              0               0              08:06                0   
2016020039              1               1              05:59                0   
2016020056              0               0              01:47                0   
2016020071              0               0              01:58                0   
...                   ...             ...                ...              ...   
2016021169              0               0              01:47                0   
2016021182              0               1              02:26                0   
2016021193              0               1              02:19                0   
2016021209              0               0              05:09                0   
2016021228              0               0              02:44                0   

           gameWinningGoals overTimeGoals shots penaltyMinutes shifts  \
2016020007                0             0     5              0     26   
2016020022                0             0     3              0     18   
2016020039                1             0     5              6     20   
2016020056                1             0     2              0     22   
2016020071                0             0     6              0     23   
...                     ...           ...   ...            ...    ...   
2016021169                0             0     2              2     20   
2016021182                0             0     3              0     22   
2016021193                0             0     2              0     22   
2016021209                0             0     3              0     21   
2016021228                0             0     5              0     20   

           timeOnIce  isHome  isWin   isOT    season  
2016020007     22:28   False  False   True  20162017  
2016020022     19:07    True   True  False  20162017  
2016020039     18:11    True   True  False  20162017  
2016020056     16:42   False   True  False  20162017  
2016020071     19:44    True  False  False  20162017  
...              ...     ...    ...    ...       ...  
2016021169     15:27   False   True  False  20162017  
2016021182     18:57   False   True  False  20162017  
2016021193     19:22    True   True  False  20162017  
2016021209     17:52   False   True  False  20162017  
2016021228     17:56    True  False  False  20162017  

[82 rows x 20 columns]

Отлично, получили информацию о 1316 играх (актуально на январь 2023 года)  
Проверили что в 2016-17 году Овечкин сыграл 82 игры

In [34]:
#сохраняем в файл pickle
total_df.to_pickle('./ovechkin.pkl')